files/lib/chef_compat/copied_from_chef/chef/dsl/declare_resource.rb

```diff
@@ -94,6 +94,10 @@ def declare_resource(type, name, created_at=nil, run_context: self.run_context,
       def build_resource(type, name, created_at=nil, run_context: self.run_context, &resource_attrs_block)
         created_at ||= caller[0]
         Thread.exclusive do
+          begin
+            require 'chef/resource_builder' unless defined?(Chef::ResourceBuilder)
+          rescue LoadError
+          end
           require "chef_compat/copied_from_chef/chef/resource_builder" unless defined?(Chef::ResourceBuilder)
```

Comment:
If we have not yet required the actual chef resource builder [when we create the class](https://github.com/chef-cookbooks/compat_resource/blob/master/files/lib/chef_compat/copied_from_chef/chef/resource_builder.rb#L25) the `ChefCompat::CopiedFromChef::Chef::ResourceBuilder` then it's super class will be Object.

And then that's a problem later [when it checks again in initialize and calls super](https://github.com/chef-cookbooks/compat_resource/blob/master/files/lib/chef_compat/copied_from_chef/chef/resource_builder.rb#L39). Because it won't be calling initialize of `Chef::ResourceBuilder` like it thinks, it will be calling the initialize of `Object` and it doesn't know what to do with those args.

This fixes the problem by making sure `Chef::ResourceBuilder` is defined (and rescuing the LoadError if it cannot be required) before defining`ChefCompat::CopiedFromChef::Chef::ResourceBuilder`

I agree that the way resource_builder is required in Chef is sloppy, but this cookbook needs to not screw with those classes if they're already there right? It's only supposed to be adding stuff to chef versions `~> 12.0` and `< 12.5`. So if in Chef 12.6.0 and 12.7.2 it replaces `Chef::ResourceBuilder` with a class that cannot be initialized, that seems like a problem. It makes it so we can't use any cookbook that uses compat_resource with those versions.

And even if they fix chef/chef#4458, are they going to yank the old gems where it did the lazy require? I doubt it.